src/core/thread/network_data_publisher.hpp

```diff
@@ -0,0 +1,374 @@
+            uint16_t            GetPort(void) const { return mPortOrSeqNumber; }
+            const Ip6::Address &GetAddress(void) const { return mAddress; }
+            void                SetAddress(const Ip6::Address &aAddress) { mAddress = aAddress; }
+
+            static Info InfoAnycast(uint8_t aSequenceNumber) { return Info(kTypeAnycast, aSequenceNumber); }
+            static Info InfoUnicast(Type aType, const Ip6::Address &aAddress, uint16_t aPort)
+            {
+                return Info(aType, aPort, &aAddress);
+            }
+
+        private:
+            Info(Type aType, uint16_t aPortOrSeqNumber, const Ip6::Address *aAddress = nullptr);
+
+            Ip6::Address mAddress;
+            uint16_t     mPortOrSeqNumber;
+            Type         mType;
+        };
+
+        Type GetType(void) const { return mInfo.GetType(); }
+        void Publish(const Info &aInfo);
+        bool Add(void);
+        bool Remove(State aNextState = kNoEntry);
+        void Process(void);
+        void CountAnycastEntries(uint8_t &aNumEntries, uint8_t &aNumPreferredEntries) const;
+        void CountUnicastEntries(uint8_t &aNumEntries, uint8_t &aNumPreferredEntries) const;
+
+        Info mInfo;
+    };
+#endif // OPENTHREAD_CONFIG_TMF_NETDATA_SERVICE_ENABLE
+
+#if OPENTHREAD_CONFIG_BORDER_ROUTER_ENABLE
+    enum : uint8_t
+    {
+        // Max number of prefix (on-mesh or external route) entries.
+        kMaxPrefixEntries = OPENTHREAD_CONFIG_NETDATA_PUBLISHER_MAX_PREFIX_ENTRIES,
+    };
+
+    class PrefixEntry : public Entry, private NonCopyable
+    {
+        friend class Entry;
+
+    public:
+        void Init(Instance &aInstance) { Entry::Init(aInstance); }
+        bool IsInUse(void) const { return GetState() != kNoEntry; }
+        bool Matches(const Ip6::Prefix &aPrefix) const { return mPrefix == aPrefix; }
+        void Publish(const OnMeshPrefixConfig &aConfig);
+        void Publish(const ExternalRouteConfig &aConfig);
+        void Unpublish(void);
+        bool HandleTimer(void) { return Entry::HandleTimer(); }
+        void HandleNotifierEvents(Events aEvents);
+
+    private:
+        enum : uint8_t
+        {
+            kDesiredNumOnMeshPrefix  = OPENTHREAD_CONFIG_NETDATA_PUBLISHER_DESIRED_NUM_ON_MESH_PREFIX_ENTRIES,
+            kDesiredNumExternalRoute = OPENTHREAD_CONFIG_NETDATA_PUBLISHER_DESIRED_NUM_EXTERNAL_ROUTE_ENTRIES,
+        };
+
+        enum Type : uint8_t
+        {
+            kTypeOnMeshPrefix,
+            kTypeExternalRoute,
+        };
+
+        bool Add(void);
+        bool Remove(State aNextState = kNoEntry);
+        void Process(void);
+        void CountOnMeshPrefixEntries(uint8_t &aNumEntries, uint8_t &aNumPreferredEntries) const;
+        void CountExternalRouteEntries(uint8_t &aNumEntries, uint8_t &aNumPreferredEntries) const;
+
+        Type        mType;
+        Ip6::Prefix mPrefix;
+        bool        mStable;
```

Comment:
Yes, we treat them separately since they do behave differently and can be used by nodes in different ways.

Practically, I expect publisher to be used with stable entries.